### PR TITLE
Improve the msp_displayport funciton

### PR DESCRIPTION
**Pull request by ligenxxxx**
1. The msp displayport function can be turned on or off by setting the port function (1<<24) of the series in the CLI. msp_displayport will occupy 1 of the 3 msp serial ports at most.
2. Modified ````UART_TX_BUFFER_SIZE```` to deal with a large amount of osd elements data suddenly generated to prevent data loss.
3. HeartBeat is removed from the msp_displayport function. And drawScreen is set to execute once every 4 drawOsdElement, the user should update the previously received osdElements to the screen when drawScreen.
4. ````OSD_ARTIFICIAL_HORIZON```` is removed under the current msp_displayport function, otherwise the system will freeze.
5. I wrote a simple test program in python on the PC, and the actual effect can be observed by connecting to the computer via serial port to USB. If someone wants to test I can send the exe to a personal email.

### CODE REVIEW

File: src/main/drivers/serial_uart.h [hunk 1]
```diff
@@ -25,6 +25,24 @@
 // Size must be a power of two due to various optimizations which use 'and' instead of 'mod'
 // Various serial routines return the buffer occupied size as uint8_t which would need to be extended in order to
 // increase size further.
+#ifdef USE_MSP_DISPLAYPORT
+#define UART1_RX_BUFFER_SIZE    256
+#define UART1_TX_BUFFER_SIZE    1280
```
Comment (DzikuVx) [inline src/main/drivers/serial_uart.h]: @ligenxxxx why 1280 and not 1024? The comment above mentions that buffer should the the power of 2

File: src/main/io/osd.c [hunk 2]
```diff
@@ -2537,7 +2537,9 @@ void osdDrawNextElement(void)
     } while(!osdDrawSingleElement(elementIndex) && index != elementIndex);
 
     // Draw artificial horizon last
+#ifndef USE_MSP_DISPLAYPORT
```
Comment (DzikuVx) [inline src/main/io/osd.c]: @ligenxxxx why this is required? I'd like to understand

File: src/main/io/osd.c [hunk 3]
```diff
@@ -3177,7 +3179,9 @@ static void osdRefresh(timeUs_t currentTimeUs)
             fullRedraw = false;
         }
         osdDrawNextElement();
+#ifndef USE_MSP_DISPLAYPORT
```
Comment (DzikuVx) [inline src/main/io/osd.c]: This disabled the `displayHeartbeat` when MSP_DISPLAYPORT is compiled in. It even does not have to be enabled to disable

File: src/main/io/osd.c (hunk 2 above)
Comment (DzikuVx) [reply src/main/io/osd.c]: And I think it should be when MSP_DISPLAYPORT is enabled and running (check at runtime), not only compiled into the code.
With current approach, if you compile with `USE_MSP_DISPLAYPORT` AHI stops working even if you do not use DISPLAYPORT

**Review by DzikuVx — CHANGES_REQUESTED** [review]

File: src/main/drivers/serial_uart.h (hunk 1 above)
Comment (ligenxxxx) [reply src/main/drivers/serial_uart.h]: I would like to modify the depth to 1024, for the current mode of transmission, the loss of a small amount of data is feasible.

File: src/main/io/osd.c (hunk 2 above)
Comment (ligenxxxx) [reply src/main/io/osd.c]: This osd elements can make the system less fluid, and I don't know why.

File: src/main/io/osd.c (hunk 3 above)
Comment (ligenxxxx) [reply src/main/io/osd.c]: I understand what you said now.

File: src/main/io/osd.c (hunk 2 above)
Comment (DzikuVx) [reply src/main/io/osd.c]: Then it should be disabled when MSP_DISPLAYPORT is configured and enabled, not only compiled

**Comment by Dronitoni** [comment]
Is this something what goes into the direction of using more OSD-Elements with Sharkbyte?

If yes, I would like to support if I can, because I'm using Sharkbyte on a 7-incher with INAV.
(Unfortunately I'm not a software guy).

If some SW needs to be tested I will support.

ligenxxxx you are from divimath isn't it?

**Review by ligenxxxx — COMMENTED** [review]
> Is this something what goes into the direction of using more OSD-Elements with Sharkbyte?
> 
> If yes, I would like to support if I can, because I'm using Sharkbyte on a 7-incher with INAV.
> (Unfortunately I'm not a software guy).
> 
> If some SW needs to be tested I will support.
> 
> ligenxxxx you are from divimath isn't it?

Yes, sharkbyte is currently developing the msp displayport function to maximize compatibility with the osd of the flight controller itself. I am not sure if the current modification is feasible, whether it will affect other functions, or whether it is worth passing, I need someone to help review it.
But currently only BTFL has a complete msp displayport function. I hope that this request can be completed before sharkbyte releases a firmware version that uses the msp displayport function.
If the sharkbyte firmware with msp displayport function is generated in the testing phase, maybe I can invite you to test it together.
Of course, the improvement of this function may also be used by DJI HD.
By the way, I am not a professional software designer, I am just interested in it, learning while practicing.

**Comment by Dronitoni** [comment]
okay, as soon there's a new SW available I will try it.
I'm flying BTFL as well in one quad, so the HW is available for both systems.

**Comment by strepto42** [comment]
Would love to see this get merged in soon, am very keen to try a full OSD with sharkbyte/inav (and I'm not the only one!) :) Keep up the great work guys!

**Comment by DzikuVx** [comment]
It's a pity that noone from the INAV team has access to the Shark Byte to be able to test it

**Comment by Dronitoni** [comment]
There's an information from "Man Cave Hobbies" in his last video, that Carl from Divimath has delivered some code to the INAV-Devs. 
Here's the link:
https://youtu.be/9JYxCNl1B_k?t=740

**Comment by ligenxxxx** [comment]
> It's a pity that noone from the INAV team has access to the Shark Byte to be able to test it

My test method is to connect the serial port to the PC, and write a test program on the PC to parse the MSP command and display the osd on the screen.
Possible problem: Due to the different osd_icons of INAV and BTFL, this may result in failure to correspond when the icon is displayed. But it does not matter, as long as someone needs it, the function will be perfected.
I have Sharkbyte's new firmware, I will test it as soon as possible and upload the test results. In addition, Fatshark will release new firmware in the near future.
Since the PR takes a long time, I can't remember the details. Give me some time to test.

**Comment by ligenxxxx** [comment]
To verify MSP_DISPLAYPORT on the PC, the steps are as follows:
1. Update the PR to FC
2. I use the serial port 4, connect the serial port 4 to the serial port tool, and connect the serial port tool to the computer.
3. Open INAV Configurator, enter the CLI interface to configure MSP_DISPLAYPORT as follows:
             serial 3 16777216 115200 115200 0 115200
             save
     16777216 is a new function I added, you can refer to the code for details.
     At this point, the serial port is already outputting MSP_displayport message.
4. On the PC side, use the program I wrote to check the display status (Ascii can only be displayed correctly because the osd icon cannot be displayed in the terminal).

msp_displayport0
msp_displayport1

**Comment by ligenxxxx** [comment]
The display in sharkbyte is the same as expected, and the icon will be garbled.
Unexpectedly, the boot screen was not cleaned, but there was no such problem in the PC test, so I think the high probability may be that there is some problem with sharkbyte itself.
I think sharkbyte should make some adjustments to adapt to INAV.
(Please ignore the poor image quality) 
新建位图图像 (2)
新建位图图像

File: src/main/fc/fc_init.c [hunk 4]
```diff
@@ -554,6 +554,12 @@ void init(void)
             osdDisplayPort = frskyOSDDisplayPortInit(osdConfig()->video_system);
         }
 #endif
+#if defined(USE_MSP_DISPLAYPORT)
+        if (!osdDisplayPort) {
+            osdDisplayPort = displayPortMspInit();
+
```
Comment (wvarty) [inline src/main/fc/fc_init.c]: This extra line of whitespace can be removed.

File: src/main/io/displayport_msp.c [hunk 5]
```diff
@@ -85,7 +82,12 @@ static int clearScreen(displayPort_t *displayPort)
 static int drawScreen(displayPort_t *displayPort)
 {
     uint8_t subcmd[] = { 4 };
-    return output(displayPort, MSP_DISPLAYPORT, subcmd, sizeof(subcmd));
+    if(DrawAllElement) {
```
Comment (wvarty) [inline src/main/io/displayport_msp.c]: There should be a space between `if` and the `(` bracket here.

File: src/main/msp/msp_serial.c [hunk 6]
```diff
@@ -37,8 +37,12 @@
 #include "msp/msp.h"
 #include "msp/msp_serial.h"
 
-static mspPort_t mspPorts[MAX_MSP_PORT_COUNT];
+#include "msp/msp_protocol.h"
 
+#include "config/parameter_group.h"
+#include "config/parameter_group_ids.h"
+static mspPort_t mspPorts[MAX_MSP_PORT_COUNT];
```
Comment (wvarty) [inline src/main/msp/msp_serial.c:45]: Might be a good idea to add a line return between the `#include` block and the static object here?

File: src/main/msp/msp_serial.c [hunk 7]
```diff
@@ -50,7 +54,28 @@ void resetMspPort(mspPort_t *mspPortToReset, serialPort_t *serialPort)
...
+        msp_displayport_index |= (1 << (portConfig->identifier));
+
+        if (serialPort) {
+            resetMspPort(mspPort, serialPort);
+            portIndex++;
+        }
+
+        portConfig = findNextSerialPortConfig(FUNCTION_MSP_DISPLAYPORT);
+    }
+
```
Comment (wvarty) [inline src/main/msp/msp_serial.c]: No need for the double line return here. Should just be one line of whitespace.

File: src/main/msp/msp_serial.c [hunk 8]
```diff
@@ -534,14 +562,24 @@ int mspSerialPush(uint8_t cmd, const uint8_t *data, int datalen)
 {
     int ret = 0;
 
+    //Do not output in climode
+    if(cliMode){
```
Comment (wvarty) [inline src/main/msp/msp_serial.c:576]: Should be spaces between `if` and `(` as well as between `)` and the curly brace `{`

File: src/main/msp/msp_serial.c [hunk 9]
```diff
@@ -534,14 +562,24 @@ int mspSerialPush(uint8_t cmd, const uint8_t *data, int datalen)
...
-        // Avoid unconnected ports (only VCP for now)
-        if (!serialIsConnected(mspPort->port)) {
+        
+        // Avoid unconnected or VCP ports
+        if (!serialIsConnected(mspPort->port) || mspPort->port->identifier == SERIAL_PORT_USB_VCP) {
+            continue;
+        }
+        
+        // only enable for msp_displayport
+        if((msp_displayport_index & (1 << mspPort->port->identifier)) == 0) {
```
Comment (wvarty) [inline src/main/msp/msp_serial.c]: Space needed here too.

File: src/main/target/MATEKF405/target.h [hunk 10]
```diff
@@ -77,6 +77,7 @@
 #define SPI2_MOSI_PIN   	    PB15
 
 #define USE_OSD
+#define USE_MSP_DISPLAYPORT
```
Comment (wvarty) [inline src/main/target/MATEKF405/target.h]: Is this `#define` needed on other targets too?

Comment (ligenxxxx) [reply src/main/target/MATEKF405/target.h]: Yes. Since I tested on the target, I only added ``#define`` to this file. I am not sure if other targets also need me to maintain.

**Comment by Dronitoni** [comment]
I created the HEX-file and flashed it to a MATEKF405.
So far so good.

But if I start the INAV-Configurator it jumps automatically into CLI.
The message says "This firmware version is not supported. 

I tried several different Configurator versions including the new 3.0, but no difference.


Version-message in CLI:
# INAV/MATEKF405 2.7.0 Apr 12 2021 / 10:57:03 (GITDIR-N)
# GCC-9.3.1 20200408 (release)


What can I do ?




[inav_2.7.0_MATEKF405.zip](https://github.com/iNavFlight/inav/files/6295709/inav_2.7.0_MATEKF405.zip)

**Comment by avsaase** [comment]
I suspect this is because of the change of version number from 2.7 to 3.0. Both the 2.6.x and the 3.0 configurators are incompatible with 2.7 firmware (because it doesn't officially exist). It's probably best to rebase, or you could change the version number to 3.0.

**Comment by wvarty** [comment]
> I created the HEX-file and flashed it to a MATEKF405.
> So far so good.
> 
> But if I start the INAV-Configurator it jumps automatically into CLI.
> The message says "This firmware version is not supported.
> 
> I tried several different Configurator versions including the new 3.0, but no difference.
> 
> Version-message in CLI:
> 
> # INAV/MATEKF405 2.7.0 Apr 12 2021 / 10:57:03 (GITDIR-N)
> # GCC-9.3.1 20200408 (release)
> What can I do ?
> 
> [inav_2.7.0_MATEKF405.zip](https://github.com/iNavFlight/inav/files/6295709/inav_2.7.0_MATEKF405.zip)

You can compile and run the Configurator from the source code, which will work with this PR. Instructions are on the readme page in the Configurator repo.

**Comment by wvarty** [comment]
**Here is a quick bench test of the build:**
https://www.youtube.com/watch?v=eC0v57oDnkY

**Some notable points:**

- Generally, it works (to an extent), well done getting this far.
- As mentioned by Ligen, the font incompatibility will need to be addressed, as the text is not usable as it currently is.
- Some icons are displayed, but then become corrupted on change (again, this is the font issue).
- The OSD is generally very slow to respond (much slower than the betaflight build.
- The artificial horizon indicator did not appear to work in iNav, even though it works in BF.

For reference, here is my OSD setup screen in iNav (for some reason my GPS plus code element is not displayed, even though it is active on the top left):
osd-settings


**My test environment:**
**Build SHA:** 94cedf8b232447aca078884d521e6067de41955e
**FC:** Matek F405 Wing
**CLI version info:**  INAV/MATEKF405SE 2.7.0 Mar 31 2021 / 11:57:55 (94cedf8b) 
**Configurator version:** 2.7.0
**Shark Byte VTX:** TX5R.1
**Shark Byte FW version:** SharkByte_24032021

**Comment by avsaase** [comment]
> (for some reason my GPS plus code element is not displayed, even though it is active on the top left)

This is probably because https://github.com/iNavFlight/inav/pull/6469 or https://github.com/iNavFlight/inav-configurator/pull/1150 is not in your firmware/configurator build.

**Comment by ligenxxxx** [comment]
> 我创建了 HEX 文件并将其闪存到 MATEKF405。
> 到目前为止还不错。
> 
> 但是，如果我启动 INAV 配置器，它会自动跳转到 CLI 中。
> 消息说："此固件版本不支持。
> 
> 我尝试了几个不同的配置器版本，包括新的3.0，但没有区别。
> 
> CLI 中的版本消息：
> 
> # 伊纳夫/马特克夫405 2.7.0 四月 12 2021 / 10：57：03 （GITDIR-N）
> # 海合会-9.3.1 20200408 （发布）
> 我能做什么？
> 
> [inav_2.7.0_马泰克夫405.zip](https://github.com/iNavFlight/inav/files/6295709/inav_2.7.0_MATEKF405.zip)

Modify the 32 lines in CmakeLists.txt in the root directory, and modify the version number to 2.6.0, and just regenerate.

**Comment by ligenxxxx** [comment]
> I suspect this is because of the change of version number from 2.7 to 3.0. Both the 2.6.x and the 3.0 configurators are incompatible with 2.7 firmware (because it doesn't officially exist). It's probably best to rebase, or you could change the version number to 3.0.

Modify the 32 lines in CmakeLists.txt in the root directory, and modify the version number to 2.6.0, and just regenerate.

**Comment by ligenxxxx** [comment]
> **Here is a quick bench test of the build:**
> https://www.youtube.com/watch?v=eC0v57oDnkY
> 
> **Some notable points:**
> 
> * Generally, it works (to an extent), well done getting this far.
> * As mentioned by Ligen, the font incompatibility will need to be addressed, as the text is not usable as it currently is.
> * Some icons are displayed, but then become corrupted on change (again, this is the font issue).
> * The OSD is generally very slow to respond (much slower than the betaflight build.
> * The artificial horizon indicator did not appear to work in iNav, even though it works in BF.
> 
> For reference, here is my OSD setup screen in iNav (for some reason my GPS plus code element is not displayed, even though it is active on the top left):
> osd-settings
> 
> **My test environment:**
> **Build SHA:** [94cedf8](https://github.com/iNavFlight/inav/commit/94cedf8b232447aca078884d521e6067de41955e)
> **FC:** Matek F405 Wing
> **CLI version info:** INAV/MATEKF405SE 2.7.0 Mar 31 2021 / 11:57:55 ([94cedf8](https://github.com/iNavFlight/inav/commit/94cedf8b232447aca078884d521e6067de41955e))
> **Configurator version:** 2.7.0
> **Shark Byte VTX:** TX5R.1
> **Shark Byte FW version:** SharkByte_24032021


    1. I currently plan to solve the font problem first.
    2. Regarding the problem of slow response, wait until the font is resolved. I think font issues have a higher priority.
    3. Opening the osd project will cause fc to freeze, which may have a bug, I temporarily blocked it when the msp_displayport function is turned on.

**Comment by Dronitoni** [comment]
> > I suspect this is because of the change of version number from 2.7 to 3.0. Both the 2.6.x and the 3.0 configurators are incompatible with 2.7 firmware (because it doesn't officially exist). It's probably best to rebase, or you could change the version number to 3.0.
> 
> Modify the 32 lines in CmakeLists.txt in the root directory, and modify the version number to 2.6.0, and just regenerate.

Okay, the HEX-file is created.
Upoladed to the MATEK F405, changed "serial 0 16777216 115200 115200 0 115200" (--> in my built serial 0 (UART1) is connected to Shakbyte VTx).

Result: no OSD on the screen
Switching the UART1 back to MSP: OSD available

Do I need to use a different FW for the Sharkbyte VTX? I'm using currently the Rev 20022021

**Comment by strepto42** [comment]
Correct, there is a special firmware for RX and I think TX. Hopefully the 
Divimath guys can supply. Thank you all for your work on this!

**Comment by ligenxxxx** [comment]
Sharkbyte FW has modified to remap osd_symbol for INAV. If you need this firmware, you can try to contact ```carl@divimath.com```, maybe you can get it. Note that this firmware is an unstable version and is only used for this test.

currently existing problems:    
         1. The remapping osd_symbol may have some inappropriate places. Someone needs to conduct a comprehensive test for this and provide suggestions for improvement.    
         2. For INAV, garbled codes sometimes appear. Currently, the problem can only be solved by restarting VTX. The reason is that sometimes VTX will parse the msp protocol incorrectly. Because INAV does not usually send the clear_screen command, the error is always displayed on the screen. (There is no problem working with BF).    
         3. The osd update speed is very slow. The cause is also related to INAV, and I am analyzing it.    

Thanks to the guys who paid for this, this will speed up the release of this feature.

**Comment by Dronitoni** [comment]
> Sharkbyte FW has modified to remap osd_symbol for INAV. If you need this firmware, you can try to contact `carl@divimath.com`, maybe you can get it. Note that this firmware is an unstable version and is only used for this test.




Okay, I've contacted carl@divimath.com directly.
Let's wait and see.

**Comment by ligenxxxx** [comment]
In the latest PR, the osd update speed has become normal.
Existing problems:
1. There will still be occasional garbled problems, but because of this modification, it will be restored soon. I stepped on the Sharkbyte firmware more likely to have a bug, and I will investigate it. BF seems to have the same problem.
2. When adding too many osd, it will freeze, I am not sure who (FC or VTX) is stuck. So when you find that the osd is stuck in time sharing, please try to remove a few osd items. I will troubleshoot this, but the garbled problem has a higher priority.

At the same time, I don't know whether this modification affects the osd of analogVTX. I did not test analog VTX.

**Comment by Dronitoni** [comment]
I received the FW from Carl and flashed it to VRX and VTX.
--> no OSD

Next step was to built the HEX-file from your repository.
https://github.com/ligenxxxx/inav.git
--> no OSD 

I used 16777216  for the serial port, alternatively used MSP, but no success.

Currently I don't know why I have no OSD on the screen.
I think I have overlooked something.

**Comment by ligenxxxx** [comment]
I will confirm on Monday.

**Comment by ligenxxxx** [comment]
> I received the FW from Carl and flashed it to VRX and VTX.
> --> no OSD
> 
> Next step was to built the HEX-file from your repository.
> https://github.com/ligenxxxx/inav.git
> --> no OSD 
> 
> I used 16777216  for the serial port, alternatively used MSP, but no success.
> 
> Currently I don't know why I have no OSD on the screen.
> I think I have overlooked something.
> 

Try to open only one osd item.  If it is still not displayed, I will check it next Monday.

**Comment by Dronitoni** [comment]
> Try to open only one osd item. If it is still not displayed, I will check it next Monday.


It makes unfortunately no difference.
--> nothing displayed

**Comment by ligenxxxx** [comment]
> > Try to open only one osd item. If it is still not displayed, I will check it next Monday.
> 
> It makes unfortunately no difference.
> --> nothing displayed

I re-checked the INAV code and the SB test version firmware, and it worked normally for me.
There are two problems here:
1. After the VTX is powered on, does the blue LED blink 8 times? If it flashes, it means that VTX and FC communicate normally.
2. Check whether the firmware of the test version works normally with BTFL. If there is no problem with the flash, it will work normally with BTFL.

**Comment by Dronitoni** [comment]
Okay, I made some more tests, here are my results:

**1. SB with test-FW (SharkByte_15042021)**

1.1 INAV with your Test-SW: 
- blue LED switches on without blink-sequence
--> no OSD Elements
- tried also with only one OSD-Element
--> no difference

1.2 Betaflight with Standard-SW:
- blue LED comes up with blink-sequence
--> no OSD Elelements
- same result with only one OSD-Element


**2. SB with released FW, dated Feb. 2021**

2.1 INAV with your Test-SW:
--> same like Pos. 1.1

2.2 INAV with Standard-SW:
- blue LED comes up with blink-sequence
--> some OSD Elements available, like previous situation

2.3 Betaflight with Standard-SW:
- blue LED comes up with blink-sequence
--> some OSD Elements available, like previous situation



What I've seen when comparing the Sharkbyte-Firmware from February and 
the Version Carl send to me (SharkByte_15042021).
The Februaray Version contains a file named **HDZERO_VA.bin**.
This file is **not** included in Carl's Version.

I don't know if this makes a difference.

**Comment by wvarty** [comment]
### Test Log #2:

**My test environment:**
**Build SHA:** be577652e6eb426ec0e1add5162e4dd2171ac8e9
**FC:** Matek F405 Wing
**CLI version info:**  INAV/MATEKF405SE 2.6.0 Apr 19 2021 / 20:45:30 (be577652)
**Configurator version:** 2.6.0
**Shark Byte VTX:** TX5R.1
**Shark Byte FW version:** SharkByte_15042021

**Another bench test video of the build:**
https://youtu.be/vx5R1Cypqxs

**More notable points:**

1. Icons for iNav are looking much better.

2. Some OSD elements will appear, then disappear, for example in the screenshot below, the throttle pos and battery cell voltage elements come and go:
OSD-differences

3. The OSD is still VERY slow to update (see video link above for throttle pos change example).

4. Some OSD elements are never displayed, for example the RC link stats, and the GPS Plus code:
OSD-missing-elements

5. The MSP displayport configuration can be overwritten when making changes on the "Ports" tab in the Configurator. For example, I flashed the FW build, applied my CLI diff to setup my wing config, then ran the serial CLI command to enable MSP displayport on UART1:
CLI-cmd
After this, I confirmed I had OSD (I did).
I then went back to the Configurator, and changed my RX UART from 2 to 3:
Ports-tab-change-config
this seems to overwrite the `serial` command applied in the previous step, as I now have no OSD:
OSD-NoOSD

6. The config for the `serial` function code `16777216` does not seem to get saved in a CLI `dump`, which means it gets cleared if you load fresh firmware, then load a diff.

**Comment by Dronitoni** [comment]
Now it's getting better.
Carl pointed me to the correct HDZERO_VA-file.

The blue LED comes up with blink-sequence,
and I have OSD Elements displayed now.



But,
basically I can confirm what wvarty is describing in his posts above.

If I change anything in the CLI, this works perfect, but as soon as I modify 
something in INAV-Configurator the OSD is screwed up.
It doesn't matter if I change something in the ports tab or in the OSD tab.

**Comment by ligenxxxx** [comment]
> 5. The MSP displayport configuration can be overwritten when making changes on the "Ports" tab in the Configurator. For example, I flashed the FW build, applied my CLI diff to setup my wing config, then ran the serial CLI command to enable MSP displayport on UART1

16777216 means that serial is designated as msp_displayport function. But there is no such option in the Ports menu of the Configurator. So when save at Ports menu, 16777216 is overwritten.
The solution I can think of is to add the MSP_DISPLAYPORT option in the peripheral and specify it as 16777216. I am not sure if I have the ability to modify it. I have never contacted the Configurator.

Try to add fewer osd items, the speed should be much faster. And add on time item to monitor that there is no stuck. I found in the test that too many osd will cause it to freeze.

**Comment by Dronitoni** [comment]
Since it's basically working now,
someone from the INAV-Configurator specialists needs to jump in.

**Comment by rquellet** [comment]
Flight test was a success! Tested on a F405-wing.I like how quickly the OSD elements update and that there are no partially rendered characters like I'd seen before. One problem is that the main OSD menu doesn't appear to work (left-rudder + up pitch).
See flight results here:
https://youtu.be/PoFqWXuSJKg

What needs to be worked on to keep this feature moving forward? Thanks guys!

**Comment by rquellet** [comment]
I conducted a second test last night with the Matek F722WPX flight controller. Video and comments on the software here:
https://youtu.be/DRB-QH13TC4

- Had to add MSP_DISPLAYPORT to F722 targets in order to enable feature
- Had to disable frsky pixel OSD on ports tab in configurator, otherwise MSP_DISPLAYPORT didn't function

**Comment by Dronitoni** [comment]
@rquellet, 
Have you seen the same behavior I have, that as soon as you change anything in the configurator, the OSD disappears?

**Comment by rquellet** [comment]
> @rquellet,
> Have you seen the same behavior I have, that as soon as you change anything in the configurator, the OSD disappears?

If I change something on the Ports tab and save, then the serial mapping is overwritten. Changes on other tabs didn't overwrite my serial mapping for MSP_DISPLAYPORT

**Comment by rquellet** [comment]
I think I've found a bug. It appears that the direction to home arrow is pointing in the opposite direction that I'd expect it to point in. In this screenshot, I'm flying away from myself, so the direction to home arrow should point down, right?
image

**Comment by ligenxxxx** [comment]
> I think I've found a bug. It appears that the direction to home arrow is pointing in the opposite direction that I'd expect it to point in. In this screenshot, I'm flying away from myself, so the direction to home arrow should point down, right?
> image

I read the source code. Due to my negligence, there is indeed a problem with the mapping of the Direction arrow Symbol. BTFL and INAV are opposite on UP and DOWN, but RIGHT and LEFT should be no problem. This problem will be solved on the sharkbyte side.

File: CMakeLists.txt [hunk 11]
```diff
@@ -29,7 +29,7 @@ else()
     include("${CMAKE_CURRENT_SOURCE_DIR}/cmake/${TOOLCHAIN}-checks.cmake")
 endif()
 
-project(INAV VERSION 2.7.0)
+project(INAV VERSION 2.6.0)
```
Comment (DzikuVx) [inline CMakeLists.txt]: The changes to this file should be reverted.

File: src/main/io/osd_common.c [hunk 12]
```diff
@@ -116,8 +116,12 @@ void osdDrawArtificialHorizon(displayPort_t *display, displayCanvas_t *canvas, c
         osdCanvasDrawArtificialHorizon(display, canvas, p, pitchAngle, rollAngle);
     } else {
 #endif
+#if defined(USE_MSP_DISPLAYPORT)
```
Comment (DzikuVx) [inline src/main/io/osd_common.c]: This is breaking change. It should depend not on build with `USE_MSP_DISPLAYPORT` but on the `FUNCTION_MSP_DISPLAYPORT` enabled on one of the ports.
Right now, it disables 
```
osdDrawPointGetGrid(&gx, &gy, display, canvas, p);
        osdGridDrawArtificialHorizon(display, gx, gy, pitchAngle, rollAngle);
```
even if displayport is not used.
It has to be changed to dynamic, not static condition

File: src/main/msp/msp_serial.c [hunk 13]
```diff
@@ -50,7 +55,27 @@ void resetMspPort(mspPort_t *mspPortToReset, serialPort_t *serialPort)
 void mspSerialAllocatePorts(void)
 {
     uint8_t portIndex = 0;
-    serialPortConfig_t *portConfig = findSerialPortConfig(FUNCTION_MSP);
+    //FUNCTION_MSP_DISPLAYPORT detect
+    serialPortConfig_t *portConfig = findSerialPortConfig(FUNCTION_MSP_DISPLAYPORT);
```
Comment (DzikuVx) [inline src/main/msp/msp_serial.c:59]: This block should be extracted into a separate function and called only when `USE_MSP_DISPLAYPORT`
Now it will run even if `USE_MSP_DISPLAYPORT ` is not defined for a target

File: src/main/io/osd.c [hunk 14]
```diff
@@ -3245,22 +3233,19 @@ void osdUpdate(timeUs_t currentTimeUs)
     }
 #endif
 
-#define DRAW_FREQ_DENOM     4
-#define STATS_FREQ_DENOM    50
+#define DRAW_FREQ_DENOM     25  // 250HZ/25 = 10HZ
+#define STATS_FREQ_DENOM    50  // 250HZ/50 = 5HZ
```
Comment (DzikuVx) [inline src/main/io/osd.c]: Why the change? What was wrong with the previous approach?

File: src/main/io/osd.c [hunk 15]
```diff
@@ -3122,15 +3121,8 @@ static void osdRefresh(timeUs_t currentTimeUs)
 {
     osdFilterData(currentTimeUs);
 
-#ifdef USE_CMS
```
Comment (DzikuVx) [inline src/main/io/osd.c]: This should not be changed. The new behavior should be added as an option, executed ONLY is displayport is in use

File: src/main/io/osd.c [hunk 16]
```diff
@@ -3172,11 +3164,7 @@ static void osdRefresh(timeUs_t currentTimeUs)
 #ifdef USE_CMS
     if (!displayIsGrabbed(osdDisplayPort)) {
         displayBeginTransaction(osdDisplayPort, DISPLAY_TRANSACTION_OPT_RESET_DRAWING);
-        if (fullRedraw) {
```
Comment (DzikuVx) [inline src/main/io/osd.c:3175]: This seems to be breaing cuyrrent analog implementation of the OSD. Displayport adaptations should be added as an option when used, not always

File: src/main/target/MATEKF405/target.h (hunk 10 above)
Comment (DzikuVx) [reply src/main/target/MATEKF405/target.h]: Then it should be added to all targets either per target or in `common.h`

**Review by DzikuVx — CHANGES_REQUESTED** [review]

**Comment by DzikuVx** [comment]
@ligenxxxx there are a series of changes that are breaking currect analog OSD implementation. Those problems have to be resolved before this can be merged

**Comment by DzikuVx** [comment]
and additionally, please resolve all the conflicts

File: src/main/io/osd.c (hunk 14 above)
Comment (ligenxxxx) [reply src/main/io/osd.c]: Too fast refresh speed will cause the serial port data volume to be too large, even exceeding 115200bps. I have previously tested that when enough osd is turned on at 10HZ refresh rate to cover the entire screen, the serial port will probably be occupied by 50%～60% at 115200bps.

File: src/main/msp/msp_serial.c (hunk 13 above)
Comment (ligenxxxx) [reply src/main/msp/msp_serial.c:59]: Since msp_displayport also has the function of msp port, I don't think it should be used as a separate function.

File: src/main/target/MATEKF405/target.h (hunk 10 above)
Comment (ligenxxxx) [reply src/main/target/MATEKF405/target.h]: ````common.h```` have added it.

File: src/main/msp/msp_serial.c (hunk 13 above)
Comment (DzikuVx) [reply src/main/msp/msp_serial.c:59]: It is a separate port function. Yes, it might have MSP in the name and use MSP, but is misleading from the code point of view

**Comment by rquellet** [comment]
Greqt to see some progress again, keep it up! Another flight test with this PR merged in https://youtu.be/tguqW82WOp4

This was with a Matek F722-mini. As I noted in the video above, I experienced issues where the flight controller crashed (I think) when configuring the FC via the configurator after enabling msp displayport. Similar behavior experienced (crash) if I powered off the Shark Byte VTX after boot up of the FC. Any ideas? Works great otherwise

**Comment by Dronitoni** [comment]
I tried to make some more tests with the updated "inav-Master" files, but it didn't work.
When building  a .hex-file for e.g. MATEKF405 I get lot's of error messages, see below:

/mnt/c/inav-master/src/utils/settings.rb:884:in `block in resolve_types': Could not resolve type for member osd_ahi_style in group PG_OSD_CONFIG (RuntimeError)
make[3]: *** [src/main/target/MATEKF405/CMakeFiles/MATEKF405.elf.dir/build.make:63: src/main/target/MATEKF405/MATEKF405/settings_generated.h] Error 1
make[2]: *** [CMakeFiles/Makefile2:30252: src/main/target/MATEKF405/CMakeFiles/MATEKF405.elf.dir/all] Error 2
make[1]: *** [CMakeFiles/Makefile2:30287: src/main/target/MATEKF405/CMakeFiles/MATEKF405.dir/rule] Error 2
make: *** [Makefile:12208: MATEKF405] Error 2


I tried also to create a .hex-file with the inav 3.0 repository, this works without any trouble.


What can be done?

**Comment by ligenxxxx** [comment]
> I tried to make some more tests with the updated/latest "inav-master" files, but it didn't work.
> When building a .hex-file for e.g. MATEKF405 I get lot's of error messages, see below:
> 
> /mnt/c/inav-master/src/utils/settings.rb:884:in `block in resolve_types': Could not resolve type for member osd_ahi_style in group PG_OSD_CONFIG (RuntimeError)
> make[3]: *** [src/main/target/MATEKF405/CMakeFiles/MATEKF405.elf.dir/build.make:63: src/main/target/MATEKF405/MATEKF405/settings_generated.h] Error 1
> make[2]: *** [CMakeFiles/Makefile2:30252: src/main/target/MATEKF405/CMakeFiles/MATEKF405.elf.dir/all] Error 2
> make[1]: *** [CMakeFiles/Makefile2:30287: src/main/target/MATEKF405/CMakeFiles/MATEKF405.dir/rule] Error 2
> make: *** [Makefile:12208: MATEKF405] Error 2
> 
> I tried also to create a .hex-file with the inav 3.0 repository, this works without any trouble.
> 
> What can be done?

There seems to be some problems, please be patient, I will re-upload the source code within this week.

**Comment by ligenxxxx** [comment]
Some conflicts are caused by the master update. I don't know how to resolve it, so I plan to resubmit the PR.

**Comment by BashkaIF** [comment]
direct me please how to make OSD work with latest INAV I built master with SOFT_I2C ащк MAMBAF405US
I have latest firmware on sharkbyte RX/TX,

**Comment by Dronitoni** [comment]
this one is closed.

You just need to go to this one:
https://github.com/ligenxxxx/inav-3.0.1#readme

Clone the data and create your *.hex file.
